src/main/java/org/wildfly/security/auth/realm/IdentitySharedExclusiveLock.java

```diff
@@ -0,0 +1,157 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2016 Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.wildfly.security.auth.realm;
+
+/**
+ * A simple shared/exclusive lock for a realm identity.
+ *
+ * @author <a href="mailto:fjuma@redhat.com">Farah Juma</a>
+ */
+public class IdentitySharedExclusiveLock {
+
+    private int sharedHoldCount;
+    private boolean isExclusiveLocked;
+    private int exclusiveRequests;
+    private Thread exclusiveThread = null;
+
+    /**
+     * Acquire the exclusive lock. An invocation of this method will block until the lock can be acquired
+     * or the invoking thread is interrupted.
+     *
+     * @return a lock object representing the newly acquired lock
+     */
+    public synchronized IdentityLock lockExclusive() {
+        boolean interrupted = false;
+        try {
+            exclusiveRequests++;
+            while ((sharedHoldCount > 0) || isExclusiveLocked) {
+                try {
+                    wait();
+                } catch (InterruptedException e) {
+                    interrupted = true;
+                }
+            }
+            isExclusiveLocked = true;
+            exclusiveRequests--;
+            exclusiveThread = Thread.currentThread();
+            return new IdentityLock(true);
+        } finally {
+            if (interrupted) {
+                Thread.currentThread().interrupt();
+            }
+        }
+
+    }
+
+    /**
+     * Acquire a shared lock. An invocation of this method will block until the lock can be acquired or
+     * the invoking thread is interrupted.
+     *
+     * @return a lock object representing the newly acquired lock
+     */
+    public synchronized IdentityLock lockShared() {
+        boolean interrupted = false;
+        try {
+            while ((isExclusiveLocked && ! isExclusiveThread(Thread.currentThread())) && (exclusiveRequests > 0)) {
+                try {
+                    wait();
+                } catch (InterruptedException e) {
+                    interrupted = true;
+                }
+            }
+            sharedHoldCount++;
+            return new IdentityLock(false);
+        } finally {
+            if (interrupted) {
+                Thread.currentThread().interrupt();
+            }
+        }
+    }
+
+    private synchronized boolean isExclusiveThread(Thread callingThread) {
```

Comment:
I think this might be problematic.  A thread may be reused any number of times (in say a thread pool) meaning that using the calling thread to determine the owner for the purposes of reentrancy won't work - or it may work sometimes and not other times, depending how how tasks might get dispatched to executor threads.

Given that IdentityLock is a succinct representation of ownership, I think that the user should use that instance to represent ownership, and to track reentrancy on their own.